server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -509,9 +511,15 @@ private void searchLeaf(LeafReaderContext ctx, Weight weight, Collector collecto
                 }
                 try {
                     bulkScorer.score(leafCollector, liveDocs);
-                } catch (CollectionTerminatedException e) {
+                } catch (CollectionTerminatedException | TimeLimitingCollector.TimeExceededException e) {
+                    leafCollector.finish();
+                    // We did not terminate collection early, rethrow the time exceeded exception
+                    // Otherwise
                     // collection was terminated prematurely
                     // continue with the following leaf
+                    if (e instanceof TimeLimitingCollector.TimeExceededException) {
```

Comment:
I thought we don't use TimeLimitingCollector anywhere and we can skip handling this exception.

Reply:
@javanna let me check. I am concerned that if we exit early for any valid reason, we should ensure we finish the leaf.